Improve .gitignore and LoadDocument and LoadFont documentation

hobyst opened the pull request:
Improves the .gitignore file to ignore all IDE and code editor setting folders regardless of their names and improves the documentation and function interface of `LoadDocument()` and `LoadFontFace()` in their file path variants regarding on how relative paths are treated.

The Doxygen comments have been written using the official Markdown compatibility spec of Doxygen, which some IDEs might have issues displaying it the right way due to not implementing Doxygen comments properly. However, by Doxygen standards they are just fine.

- The docstrings being displayed in Doxygen:
![imagen](https://user-images.githubusercontent.com/48522412/179748561-2c7607fa-6bc4-4b87-81a9-3f57460aaf5a.png)
![imagen](https://user-images.githubusercontent.com/48522412/179748597-0e405d21-d92b-44eb-bd0b-64a52db91d2a.png)

- The docstrings being displayed in Visual Studio Code:
![imagen](https://user-images.githubusercontent.com/48522412/179748954-6368e857-f7df-4285-926d-008ba03f22f4.png)
![imagen](https://user-images.githubusercontent.com/48522412/179748927-579009a9-36c5-4a42-b0b0-76777a35bde1.png)

hobyst commented:
Emscripten build fails due to compiler warning as error policy in a header file not affected in this pull request:
```text
/home/runner/work/RmlUi/RmlUi/Source/Debugger/../../Include/RmlUi/Core/../Config/../Core/Containers/robin_hood.h:1430:33: error: builtin __has_trivial_copy is deprecated; use __is_trivially_constructible instead [-Werror,-Wdeprecated-builtins]
        Cloner<Table, IsFlat && ROBIN_HOOD_IS_TRIVIALLY_COPYABLE(Node)>()(o, *this);
```

hobyst commented:
> However, the `Load...` comments are only true if you use the shell from the samples and its platform extensions + custom file interface. These are really only intended to make the included samples work (to find the assets in various situations) and should not be used by clients. Instead, the default file interface will just forward the argument to `fopen`.

Okay so I have delete the whole mention of the search in parent folders as that's just done for the samples right?

> Just as a side note on the doxygen comments. We don't actually compile our documentation with doxygen so we mostly care for how it looks visually in raw text and secondly for IDEs. For this reason I'd discourage heavy use of markdown but in this case it looks fine though, albeit a bit on the verbose side. For more detailed behavioral comments I'd encourage submitting a PR on the documentation repository instead, there's a handy "edit on github" link on every page.

If the answer to the question I asked above is affirmative, then deleting the mention of how samples look for their assets will be enough for the comments to look fine again in IDEs and they won't be verbose anymore.

As a sidenote, having some way to build the Doxygen documentation is useful to check the entire API without necessarily having to manually document every single function in the documentation repository. I could contribute the Doxygen file I used to generate the documentation for other devs wanting to check the API reference by just running `doxygen Doxyfile` in the root folder of the repository.

mikke89 commented:
> Okay so I have delete the whole mention of the search in parent folders as that's just done for the samples right?

That's right!

> If the answer to the question I asked above is affirmative, then deleting the mention of how samples look for their assets will be enough for the comments to look fine again in IDEs and they won't be verbose anymore.

Yup, sounds good!

> As a sidenote, having some way to build the Doxygen documentation is useful to check the entire API without necessarily having to manually document every single function in the documentation repository. In another PR, I could contribute the Doxygen file I used to generate the documentation in for other devs wanting to check the API reference by just running `doxygen Doxyfile` in the root folder of the repository.

I don't really find Doxygen documentation very helpful myself, but if you do then that is all good. As long as the Doxygen file does not require any maintenance I'm happy to take a PR.

hobyst commented:
> I don't really find Doxygen documentation very helpful myself, but if you do then that is all good. As long as the Doxygen file does not require any maintenance I'm happy to take a PR.

The Doxygen would only need maintenance if you were to change the folders used to store the includes and the source code files of RmlUi in the top level directory, as I'm referencing `Backends`, `Include`, `Samples` and `Source` as `INPUT` directories for Doxygen. But since I enabled `RECURSIVE`, any folder structure changes inside of those will not need to be reflected inside the Doxyfile.

I just changed the commits and deleted the mention of the assets search method used by samples in the comments btw.

mikke89 commented on Include/RmlUi/Core/Context.h:
```diff
@@ -92,7 +92,8 @@ class RMLUICORE_API Context : public ScriptInterface
 	ElementDocument* CreateDocument(const String& instancer_name = "body");
 	/// Load a document into the context.
-	/// @param[in] document_path The path to the document to load.
+	/// @param[in] document_path The path to the document to load. This path can be either absolute (full path) or relative to the working directory
+	/// used to run the executable that utilizes RmlUi.
```
```suggestion
	/// @param[in] document_path The path to the document to load. The path is passed directly to the file interface which is used to load the file.
	/// The default file interface accepts both absolute paths and paths relative to the working directory.
```

mikke89 commented on Include/RmlUi/Core/Core.h:
```diff
@@ -117,12 +117,13 @@ RMLUICORE_API Context* GetContext(int index);
 
 /// Adds a new font face to the font engine. The face's family, style and weight will be determined from the face itself.
-/// @param[in] file_name The file to load the face from.
+/// @param[in] file_path The path to the file to load the face from. This path can be either absolute (full path) or relative to the working directory
+/// used to run the executable that utilizes RmlUi.
```
```suggestion
/// @param[in] file_path The path to the file to load the face from. The path is passed directly to the file interface which is used to load the file.
/// The default file interface accepts both absolute paths and paths relative to the working directory.
```

mikke89 reviewed (commented):
Thanks for the update. I think the comments can be confusing when using a custom file interface. I made some suggestions which I believe better describes the usage.

mikke89 commented:
> The Doxygen would only need maintenance if you were to change the RmlUi version number (if you want it to appear in the docs) or the folders used to store the includes and the source code files of RmlUi in the top level directory, as I'm referencing `Backends`, `Include`, `Samples` and `Source` as `INPUT` directories for Doxygen. But since I enabled `RECURSIVE`, any folder structure changes inside of those will not need to be reflected inside the Doxyfile.

Alright, that sounds good to me.

> Also, since Doxygen supports Markdown it would be possible to merge the currently existing documentation files along with everything that Doxygen generates from parsing the source code into a single source for both manually-written docs and the automated API reference. Of course, it would have to be investigated up to what exent such a merge is possible, specially on if it would be possible to port the current theming of the documentation into Doxygen. But since all the documentation Markdown files are in another repository, it might be better just to leave it as it is and use Doxygen just purely for the API reference.

Sounds like a huge undertaking. I haven't personally ever been impressed by auto-generated documentation and particularly so by Doxygen. In my opinion our current manually curated documentation is a much better approach.

hobyst commented:
Just amended my last commit with your suggested changes @mikke89.

mikke89 commented:
Perfect, thank you!
